kernel/src/util/iovec.rs

```diff
@@ -126,3 +136,195 @@ pub fn copy_iovs_from_user(start_addr: Vaddr, count: usize) -> Result<Box<[IoVec
 
     Ok(io_vecs.into_boxed_slice())
 }
+
+/// The util function for create [`VmReader`]/[`VmWriter`]s.
+fn copy_iovs_and_convert<'a, T: 'a>(
+    ctx: &'a Context,
+    start_addr: Vaddr,
+    count: usize,
+    convert_iovec: impl Fn(&IoVec, &'a VmSpace) -> Result<T>,
+) -> Result<Box<[T]>> {
+    let vm_space = ctx.process.root_vmar().vm_space();
+
+    let mut v = Vec::with_capacity(count);
+    for idx in 0..count {
+        let iov = {
+            let addr = start_addr + idx * core::mem::size_of::<UserIoVec>();
+            let uiov: UserIoVec = vm_space
+                .reader(addr, core::mem::size_of::<UserIoVec>())?
+                .read_val()?;
+            IoVec::try_from(uiov)?
+        };
+
+        if iov.is_empty() {
+            continue;
+        }
+
+        let converted = convert_iovec(&iov, vm_space)?;
+        v.push(converted)
+    }
+
+    Ok(v.into_boxed_slice())
+}
+
+/// A collection of [`VmReader`]s.
+///
+/// Such readers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecReader<'a>(Box<[VmReader<'a>]>);
+
+/// A collection of [`VmWriter`]s.
+///
+/// Such writers are built from user-provided buffer, so it's always fallible.
+pub struct IoVecWriter<'a>(Box<[VmWriter<'a>]>);
+
+impl<'a> IoVecReader<'a> {
+    /// Creates a new `IoVecReader` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let readers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::reader)?;
+        Ok(Self(readers))
+    }
+}
+
+impl<'a> IoVecWriter<'a> {
+    /// Creates a new `IoVecWriter` from user-provided io vec buffer.
+    pub fn copy_from_user(ctx: &'a Context<'a>, start_addr: Vaddr, count: usize) -> Result<Self> {
+        let writers = copy_iovs_and_convert(ctx, start_addr, count, IoVec::writer)?;
+        Ok(Self(writers))
+    }
+
+    /// Returns all writers
+    pub fn writers(&self) -> &[VmWriter] {
+        &self.0
+    }
+
+    /// Returns all mutable writers
+    pub fn writers_mut(&'a mut self) -> &mut [VmWriter] {
+        &mut self.0
+    }
+}
+
+/// Trait defining the read behavior for a collection of [`VmReader`]s.
+pub trait IoVecRead {
```

Comment:
From https://github.com/asterinas/asterinas/pull/1294#discussion_r1749622164 (in case the question is accidentally missed):
> As an alternative, have you considered/what are your feelings about implementing `IoVecRead` directly for `[VmReader]`?

I'm asking this because it has two advantages:
 - We don't have to implement the trait twice (once for `IoVecReader` and once for `VmReader`).
 - We can avoid `dyn IoVecReader` (and thus avoid the overhead of indirect function calls, i.e. calling the function pointer from its vtable).

But of course this does not seem very elegant because it somehow exposes the implementation details of `IoVecReader`.

I'm just asking for your opinion, if any. The current implementation is perfectly acceptable to me. So if you don't think there are enough benefits, maybe keeping the current implementation is good enough.

Reply:
Sorry, I did't get the meaning of your original comment correctly. So I just make effort to avoid redundant heap allocation.

Do you mean something like the following?
```rust
impl IoVecRead for &mut [VmReader]
```

I have thought of this option but it is not compatible with current APIs `FileLike::read` and `FileLike::write`, which I don't want to touch in this PR, since it will change many code in fs.

Currently, `FileLike::read` accepts a `&mut VmWriter` as parameter.
```rust
fn read(&self, writer: &mut VmWriter) -> Result<usize> 
```
However, we can't convert a `&mut VmWriter` to a `&mut [VmWriter]`. Since the network API can be called by both `FileLike::read` and `Socket::recvmsg`, the underlying API mush be compatible for both `&mut VmWriter` and `&mut [VmWriter]`. That's why I choose to implement `IoVecWrite` for both `VmWriter` and `IoVecWriter`(or [VmWriter]).

But the `FileLike` trait will be refactored in the future to accept `&mut [VmWriter]` instead of `&mut VmWriter`.  Then I think your proposed solution is better. I think the two advantages you mentioned make sense to me.